test/c/union_variant_names.sail

```diff
@@ -0,0 +1,32 @@
+default Order dec
+
+$include <prelude.sail>
+$include <string.sail>
+
+// Ensure that unions with variants that match the union name work.
+// In the past the generated C code was not compatible with C++ due
+// to different rules around anonymous unions.
+
+union n0 = {
+    n0 : unit,
+}
+
+union n1 = {
+    n2 : unit,
+    n1 : unit,
+}
+
+// newtype is syntactic sugar for a single entry union.
+newtype n3 = n3 : unit
+
+enum n4 = { n4 }
+
+function main() -> unit = {
+  // Prevent dead-code elimination of the types.
+  // This is not very robust but it works for now.
```

Comment:
If you have a more robust suggestion for this that would be good. In other languages I normally make the value depend on argc/v, or use a blackbox function, but I don't think Sail has either.

Reply:
A blackbox function wouldn't be a bad idea, but this is fine for now

Reply:
Something like:
```
function blackbox() -> int = {
    signed(undefined : bits(5))
}
```
should work as a blackbox function